pkg/liqonet/ipam/ipam.go

```diff
@@ -1181,11 +1182,43 @@ func (liqoIPAM *IPAM) SetServiceCIDR(serviceCIDR string) error {
 	}
 	// Acquire Service CIDR
 	if err := liqoIPAM.AcquireReservedSubnet(serviceCIDR); err != nil {
-		return fmt.Errorf("cannot acquire ServiceCIDR:%w", err)
+		return fmt.Errorf("cannot acquire ServiceCIDR: %w", err)
 	}
 	// Update Service CIDR
 	if err := liqoIPAM.ipamStorage.updateServiceCIDR(serviceCIDR); err != nil {
-		return fmt.Errorf("cannot set ServiceCIDR:%w", err)
+		return fmt.Errorf("cannot set ServiceCIDR: %w", err)
+	}
+	return nil
+}
+
+// SetReservedSubnets acquires and/or frees the reserved networks.
+func (liqoIPAM *IPAM) SetReservedSubnets(subnets []string) error {
+	reserved := liqoIPAM.ipamStorage.getReservedSubnets()
+
+	// Free all the reserved networks not needed anymore.
+	for _, r := range reserved {
+		if !slice.ContainsString(subnets, r) {
+			klog.Infof("freeing old reserved subnet %s", r)
+			if err := liqoIPAM.FreeReservedSubnet(r); err != nil {
+				return fmt.Errorf("una error occurred while freeing reserved subnet {%s}: %w", r, err)
+			}
+			if err := liqoIPAM.ipamStorage.updateReservedSubnets(r, updateOpRemove); err != nil {
+				return err
+			}
+		}
+	}
+
+	// Reserve the newly added subnets.
+	for _, s := range subnets {
+		if !slice.ContainsString(reserved, s) {
+			klog.Infof("acquiring reserved subnet %s", s)
+			if err := liqoIPAM.AcquireReservedSubnet(s); err != nil {
+				return fmt.Errorf("an error occurred while reserving subnet {%s}: %w", s, err)
+			}
+			if err := liqoIPAM.ipamStorage.updateReservedSubnets(s, updateOpAdd); err != nil {
+				return err
```

Comment:
IIUC, here (and less likely above), there is still the possibility of ending up with a broken scenario if this update failed, since during the next attempt the subnet would already be acquired and would then trigger a failure. Do not know if there is any solution for this problem (might free, but then also that could fail).

Reply:
A solution could be:

1. Check if the subnet to be reserved could be reserved by checking the existing acquired prefixes. If it fails return;
2. Add the subnet to the reserved list; 
3. Acquire subnet. If it fails return;
4. For al the existing subnets in the reserved list make sure that it has been acquired by checking if the prefix exists. If it does not exist then reserve the subnet.

What do you think?

Reply:
Seems fine to me